wgengine/netstack/netstack.go

```diff
@@ -891,7 +891,46 @@ func (ns *Impl) userPing(dstIP netip.Addr, pingResPkt []byte) {
 	switch runtime.GOOS {
 	case "windows":
 		err = exec.Command("ping", "-n", "1", "-w", "3000", dstIP.String()).Run()
-	case "darwin", "freebsd":
+	case "ios", "darwin":
+		_, payload := packet.ICMPEchoPayload(nil)
+
+		ifState := ns.dialer.NetMon().InterfaceState()
+		defaultIfIPs := ifState.InterfaceIPs[ifState.DefaultRouteInterface]
+		var srcIP netip.Addr
+		for _, pfx := range defaultIfIPs {
+			addr := pfx.Addr()
+			if dstIP.Is4() && addr.Is4() || dstIP.Is6() && addr.Is6() {
+				srcIP = addr
+				break
+			}
+		}
+
+		var icmph packet.Header
+		if dstIP.Is6() {
+			icmph = packet.ICMP6Header{
+				IP6Header: packet.IP6Header{
+					IPProto: ipproto.ICMPv6,
+					Src:     srcIP,
+					Dst:     dstIP,
+				},
+				Type: packet.ICMP6EchoRequest,
+				Code: packet.ICMP6NoCode,
+			}
+		} else {
+			icmph = packet.ICMP4Header{
+				IP4Header: packet.IP4Header{
+					IPProto: ipproto.ICMPv4,
+					Src:     srcIP,
+					Dst:     dstIP,
+				},
+				Type: packet.ICMP4EchoRequest,
+				Code: packet.ICMP4NoCode,
+			}
+		}
+
+		icmpPing := packet.Generate(icmph, payload)
+		err = ns.tundev.InjectOutbound(icmpPing)
```

Comment:
you're sending it off but not waiting for the reply too?

Reply:
Fixed